frame/vesting/src/lib.rs

```diff
@@ -0,0 +1,610 @@
+};
+use frame_system::{self as system, ensure_signed};
+
+type BalanceOf<T> = <<T as Trait>::Currency as Currency<<T as frame_system::Trait>::AccountId>>::Balance;
+
+pub trait Trait: frame_system::Trait {
+	/// The overarching event type.
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+
+	/// The currency trait.
+	type Currency: LockableCurrency<Self::AccountId>;
+
+	/// Convert the block number into a balance.
+	type BlockNumberToBalance: Convert<Self::BlockNumber, BalanceOf<Self>>;
+}
+
+const VESTING_ID: LockIdentifier = *b"vesting ";
+
+/// Struct to encode the vesting schedule of an individual account.
+#[derive(Encode, Decode, Copy, Clone, PartialEq, Eq, RuntimeDebug)]
+pub struct VestingInfo<Balance, BlockNumber> {
+	/// Locked amount at genesis.
+	pub locked: Balance,
+	/// Amount that gets unlocked every block after `starting_block`.
+	pub per_block: Balance,
+	/// Starting block for unlocking(vesting).
+	pub starting_block: BlockNumber,
+}
+
+impl<
+	Balance: SimpleArithmetic + Copy,
+	BlockNumber: SimpleArithmetic + Copy,
+> VestingInfo<Balance, BlockNumber> {
+	/// Amount locked at block `n`.
+	pub fn locked_at<
+		BlockNumberToBalance: Convert<BlockNumber, Balance>
+	>(&self, n: BlockNumber) -> Balance {
+		// Number of blocks that count toward vesting
+		// Saturating to 0 when n < starting_block
+		let vested_block_count = n.saturating_sub(self.starting_block);
+		let vested_block_count = BlockNumberToBalance::convert(vested_block_count);
+		// Return amount that is still locked in vesting
+		let maybe_balance = vested_block_count.checked_mul(&self.per_block);
+		if let Some(balance) = maybe_balance {
+			self.locked.saturating_sub(balance)
+		} else {
+			Zero::zero()
+		}
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Vesting {
+		/// Information regarding the vesting of a given account.
+		pub Vesting get(fn vesting):
+			map hasher(blake2_256) T::AccountId => Option<VestingInfo<BalanceOf<T>, T::BlockNumber>>;
+	}
+	add_extra_genesis {
+		config(vesting): Vec<(T::AccountId, T::BlockNumber, T::BlockNumber, BalanceOf<T>)>;
+		build(|config: &GenesisConfig<T>| {
+			// Generate initial vesting configuration
+			// * who - Account which we are generating vesting configuration for
+			// * begin - Block when the account will start to vest
+			// * length - Number of blocks from `begin` until fully vested
+			// * liquid - Number of units which can be spent before vesting begins
+			for &(ref who, begin, length, liquid) in config.vesting.iter() {
+				let balance = T::Currency::free_balance(who);
+				assert!(!balance.is_zero(), "Currencies must be init'd before vesting");
+				// Total genesis `balance` minus `liquid` equals funds locked for vesting
+				let locked = balance.saturating_sub(liquid);
+				let length_as_balance = T::BlockNumberToBalance::convert(length);
+				let per_block = locked / length_as_balance.max(sp_runtime::traits::One::one());
```

Comment:
```suggestion
				let per_block = (locked / length_as_balance).max(sp_runtime::traits::One::one());
```

Looks like we are missing parenthesis here. We want to ensure `per_block` is at least 1.

i.e. 100 balance locked over 200 blocks, will have `per_block = 1`, which actually fully vest in 100 blocks.

Reply:
Should probably add a test here for this.

Reply:
![image](https://user-images.githubusercontent.com/138296/73532739-3a386d80-4415-11ea-8ec0-df5db34af26c.png)

The code is unchanged from before. The `.max(1)` is to ensure there's never a divide by zero.

Reply:
For any sensible chain, and certainly the chains this is designed for, the average balance released per block is greater than *the smallest possible unit*, so this won't be an issue. In any case it's not something that needs fixing in this PR.

Reply:
If you want, feel free to create an issue.